File: NeetCode/Graph/Verify_Alien_Language/verify__alien__language.py
from typing import Optional, List
import heapq
from collections import defaultdict


class Solution:
    def isAlienSorted(self, words: List[str], order: str) -> bool:

        alien = {ch: index for index, ch in enumerate(order)}

        for index in range(len(words) - 1):
            first = words[index]
            second = words[index + 1]

            match = True
            ptr = 0
            while ptr < len(first) and ptr < len(second) and match:
                if alien[first[ptr]] > alien[second[ptr]]:
                    return False
                if alien[first[ptr]] < alien[second[ptr]]:
                    match = False
                ptr += 1

            if match and len(second) < len(first):
                return False

        return True

        pass


if __name__ == "__main__":
    sol = Solution()
    # words = ["dag", "disk", "dog"]
    # order = "hlabcdefgijkmnopqrstuvwxyz"

    # print(sol.isAlienSorted(words, order))

    words = ["neetcode", "neet"]
    order = "worldabcefghijkmnpqstuvxyz"

    print(sol.isAlienSorted(words, order))
    print("Running Solution...")


""" 
class Solution:
    def isAlienSorted(self, words: List[str], order: str) -> bool:

        alien = {ch: index for index, ch in enumerate(order)}

        for index in range(len(words) - 1):
            first = words[index]
            second = words[index + 1]

            ptr = 0
            length = min(len(first), len(second))
            diff = False
            for idx in range(length):

                if alien[first[idx]] > alien[second[idx]] and not diff:
                    return False
                if first[idx] != second[idx]:
                    diff = True
            if not diff and len(second) < len(first):
                return False

        return True
"""

File: NeetCode/Graph/Verify_Alien_Language/Verify_Alien_Language.cpp
// Verify_Alien_Language.cpp
//
// Compile locally with:
// clang++ Verify_Alien_Language.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    bool isAlienSorted(vector<string> &words, string order)
    {
        unordered_map<char, int> alien;
        for (int index = 0; index < order.size(); index++)
        {
            alien[order[index]] = index;
        }

        int ptr = 0;
        bool match = true;

        for (int w = 0; w < words.size() - 1; w++)
        {
            string word1 = words[w];
            string word2 = words[w + 1];

            ptr = 0;
            match = true;

            while (ptr < word1.size() && ptr < word2.size() && match)
            {
                if (alien[word1[ptr]] > alien[word2[ptr]])
                {
                    return false;
                }
                else if (alien[word1[ptr]] < alien[word2[ptr]])
                {
                    match = false;
                }

                ptr++;
            }
            if (match && word2.size() < word1.size())
            {
                return false;
            }
        }
        return true;
    }
};

int main()
{
    cout << "Running Verify_Alien_Language.cpp..." << endl;

    Solution sol;

    // TODO:
    // Add local test calls here

    return 0;
}
